debian/anim-flame.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH ANIM-FLAME 1 "June 27, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
anim-flame \- part of electricsheep, frame renderer
.SH SYNOPSIS
.B anim-flame
.RI [ options ]
.SH DESCRIPTION
This program is a part of the 
.B electricsheep(1)
package and should not be used stand-alone.
.B electricsheep(1)
uses this program to render frames for the animation.
.SH OPTIONS
This program takes arguments from
.B electricsheep(1)
.SH AUTHOR
This manual page was written by Sander Smeenk <ssmeenk@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
electricsheep (3.0.2-git20191005-1ppa1~hirsute) hirsute; urgency=low

  * Rebuild for Ubuntu/Hirsute (21.04)
  * need to add LDFLAGS+=-lpthread (Autoconf fails to detect that)
  * can also be built with libwxgtk3.0-gtk3-dev (name change in Ubuntu)
  * Note: requires flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Fri, 23 Apr 2021 18:23:55 +0100

electricsheep (3.0.2-git20191005-1ppa2~focal) focal; urgency=low

  * Rebuild for Ubuntu/Focal (20.04 LTS)
  * can also be built with libwxgtk3.0-gtk3-dev (name change in Ubuntu)
  * Note: requires flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Wed, 24 Jun 2020 23:01:17 +0000

electricsheep (3.0.2-git20191005-1+buster) Lumiera-buster-tools; urgency=low

  * Build for Debian/Buster(stable).
  * Note: requires a port of glee-dev (dropped from Debian)

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 06 Jan 2020 04:47:23 +0100

electricsheep (3.0.2-git20191005-1ppa1~bionic2) bionic; urgency=low

  * Rebuild for Ubuntu/Bionic (18.04 LTS)
  * Note: requires flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Thu, 24 Oct 2019 16:33:33 +0200

electricsheep (3.0.2-git20191005-1~stretch) Lumiera-stretch-tools; urgency=low

  * Include upstream fixes on top of release 3.0.4 (2019-10-05)
  * Linux Version bump partially applied upstream
  * there are no further changes to flam3 upstream, so latest version is OK
  * build against lib versions from the Linux distro where applicable

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 21 Oct 2019 05:37:57 +0200

electricsheep (3.0.2-git20180325-1ppa1~bionic1) bionic; urgency=low

  * Rebuild for Ubuntu/Bionic (18.04 LTS)
  * Note: needs flam 3.1.1

 -- Hermann Vosseler <deb@ichthyostega.de>  Sat, 19 May 2018 02:55:09 +0200

electricsheep (3.0.2-git20180325-1~stretch) Lumiera-stretch-tools; urgency=low

  * New upstream release 3.0.2 (2018-03-25)
  * require newest flam3 release.
  * update to build agaist libjpeg9
  * bump Debian standards version
  * update the installed desktop files
  * fix some FFmpeg deprecation warnings
  * build against lib versions from the Linux distro

 -- Hermann Vosseler <deb@ichthyostega.de>  Thu, 17 May 2018 17:50:05 +0200

electricsheep (2.7~b33-svn20150718-2ppa1~trusty) trusty; urgency=low

  * Rebuild for Ubuntu trusty (14.04 LTS)

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 10 Aug 2015 01:40:20 +0200

electricsheep (2.7~b33-svn20150718-2) Lumera-jessie-tools; urgency=low

  * fix package definition to also work on Ubuntu

 -- Hermann Vosseler <deb@ichthyostega.de>  Mon, 10 Aug 2015 01:32:23 +0200

electricsheep (2.7~b33-svn20150718-1) lumiera-jessie-tools; urgency=low

  * New upstream snapshot from 2015-07-18
  * Start over with a 'all default automatic' debian package based on CDBS
  * Build for use on Debian/Jessie

 -- Hermann Vosseler <deb@ichthyostega.de>  Sun, 09 Aug 2015 16:56:33 +0200

electricsheep (2.7~b12+svn20091224-1.1) unstable; urgency=low

  * Non-maintainer upload.
  * Patch from Ubuntu by Fabrice Coutadeur to fix FTBFS. (Closes: #621985)
    * fix_ftbfs_libav_0.7.patch: fix deprecated variable to fix a FTBFS with
      libav 0.7
    * fix_link_as-needed.patch: fix Makefile.am and Makefile.in to avoid a FTBFS
      when as-needed is used with undefined gtk functions (LP: #770773) by moving
      libs from LDFLAGS to LDADD

 -- Konstantinos Margaritis <markos@debian.org>  Fri, 03 Feb 2012 11:50:38 +0200

electricsheep (2.7~b12+svn20091224-1) unstable; urgency=low

  * New upstream snapshot (Closes: #562230)

 -- Roberto C. Sanchez <roberto@connexer.com>  Thu, 31 Dec 2009 14:59:07 -0500

electricsheep (2.7~b12+svn20090708-2) unstable; urgency=low

  * Remove contrib script (esheep-mplayer.sh) from previous Debian
    package. (Closes: #546297)
  * debian/control:
    + Add Homepage field
    + Add dependency on mplayer
  * electricsheep.desktop.gnome: Remove obsolete --zoom option
  * electricsheep.desktop: Remove obsolete --zoom option

 -- Roberto C. Sanchez <roberto@connexer.com>  Wed, 30 Dec 2009 21:41:53 -0500

electricsheep (2.7~b12+svn20090708-1) unstable; urgency=low

  * New upstream release. (Closes: #533652)
  * New maintainer.
  * Move to dehelper compatibility level 5.
  * Update Build-Depends to remove obsolete package. (Closes: #515464)
  * Remove bashism from debian/rules. (Closes: #535377)
  * Remove unneeded dpatch patches.
  * Bump Standards-Version to 3.8.3 (no changes).
  * Clean up *.desktop files.
  * Add empty watch file explaining upstream's "release" policy.
  * Update debian/copyright to reflect new year and new maintainer.
  * Clean up upstream's man page (properly escape hyphens).

 -- Roberto C. Sanchez <roberto@connexer.com>  Mon, 24 Aug 2009 16:29:53 -0400

electricsheep (2.6.8-9) unstable; urgency=low

  [ Yves-Alexis Perez ]
  * link electricsheep to the correct location. (closes: #467441)

  [ Alexis Sukrieh ]
  * Standard Revision: 3.7.3
  * Changed Apps to Applications in debian/menu (fixed a lintian warning).

 -- Alexis Sukrieh <sukria@debian.org>  Tue, 11 Mar 2008 17:30:22 +0100

electricsheep (2.6.8-8) unstable; urgency=low

  * Added a debconf-2.0 alternate to dependencies (closes: #448477)

 -- Alexis Sukrieh <sukria@debian.org>  Mon, 19 Nov 2007 16:34:53 +0100

electricsheep (2.6.8-7) unstable; urgency=low

  * Summer cleanup of the package, install .desktop files in the appropriate
    places.
    (closes: #432782)
  * Autotools cleanup.

 -- Alexis Sukrieh <sukria@debian.org>  Fri, 13 Jul 2007 15:50:13 +0200

electricsheep (2.6.8-6) unstable; urgency=low

  * Compliant with gnome-screensaver
    + provides a valid electricsheep.desktop file in
    `/usr/share/gnome-screensaver/themes'.
    (closes: #425297)

 -- Alexis Sukrieh <sukria@debian.org>  Wed, 30 May 2007 13:37:45 +0200

electricsheep (2.6.8-5) unstable; urgency=low

  * Removed the debian revision from the version string so sheeps can be
    downloaded from the server.
    (closes: #391990)


flame/anim-flame.c
/*
    flame - cosmic recursive fractal flames
    Copyright (C) 1992-2003  Scott Draves <source@flam3.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


static char *anim_flame_c_id =
"@(#) $Id: anim-flame.c,v 1.16 2004/03/04 06:35:17 spotspot Exp $";

#include "rect.h"
#include "jpeg.h"
#include <string.h>

/*
  usage is:
  env prefix=default begin=0 end=0 dtime=1 fields=0
      in=filename out=filename
      qs=v ss=v <exe>
      [ < in ]

*/

#define argi(s,d)   ((ai = getenv(s)) ? atoi(ai) : (d))
#define argf(s,d)   ((ai = getenv(s)) ? atof(ai) : (d))
#define args(s,d)   ((ai = getenv(s)) ? ai : (d))


int main(int argc, char **argv) {
  char *ai, *fname;
  char *prefix = args("prefix", "");
  int first_frame = argi("begin", 0);
  int last_frame = argi("end", 0);
  int frame_time = argi("time", 0);
  int dtime = argi("dtime", 1);
  int do_fields = argi("fields", 0);
  double qs = argf("qs", 1.0);
  double ss = argf("ss", 1.0);
  char *format = getenv("format");
  int time, channels;
  unsigned char *image;
  control_point *cps;
  int i, ncps = 0;
  char *inf = getenv("in");
  double pixel_aspect = argf("pixel_aspect", 1.0);
  FILE *in;
  frame_spec f;

  if (getenv("frame")) {
    if (getenv("time")) {
      fprintf(stderr, "cannot specify both time and frame.\n");
      exit(1);
    }
    if (getenv("begin") || getenv("end")) {
      fprintf(stderr, "cannot specify both time and begin or end.\n");
      exit(1);
    }
    first_frame = last_frame = atoi(getenv("frame"));
  }

  if (getenv("time")) {
    if (getenv("begin") || getenv("end")) {
      fprintf(stderr, "cannot specify both time and begin or end.\n");
      exit(1);
    }
    first_frame = last_frame = frame_time;
  }

  if (NULL == format) format = "jpg";
  if (strcmp(format, "jpg") &&
      strcmp(format, "ppm") &&
      strcmp(format, "png")) {
      fprintf(stderr, "format must be either jpg, ppm, or png, not %s.\n", format);
      exit(1);
  }

  fname = malloc(strlen(prefix) + 20);

   if (pixel_aspect <= 0.0) {
     fprintf(stderr, "pixel aspect ratio must be positive, not %g.\n",
	     pixel_aspect);
     exit(1);
   }

  if (inf)
    in = fopen(inf, "rb");
  else
    in = stdin;
  if (NULL == in) {
    perror(inf);
    exit(1);
  }

  cps = parse_control_points_from_file(in, &ncps);
  if (NULL == cps) {
    exit(1);
  }
  if (0 == ncps) {
    fprintf(stderr, "error: no control points.\n");
    exit(1);
  }
  for (i = 0; i < ncps; i++) {
    cps[i].sample_density *= qs;
    cps[i].height = (int)(cps[i].height * ss);
    cps[i].width = (int)(cps[i].width * ss);
    cps[i].pixels_per_unit *= ss;
  }
  if (!getenv("time") && !getenv("frame")) {
    if (!getenv("begin")) {
      first_frame = (int) cps[0].time;
    }
    if (!getenv("end")) {
      last_frame = (int) cps[ncps-1].time - 1;
      if (last_frame < first_frame) last_frame = first_frame;
    }
  }
  channels = strcmp(format, "png") ? 3 : 4;
  f.temporal_filter_radius = argf("blur", 0.5);
  f.pixel_aspect_ratio = pixel_aspect;
  f.cps = cps;
  f.ncps = ncps;
  image = (unsigned char *) malloc(channels * cps[0].width * cps[0].height);
   
  for (time = first_frame; time <= last_frame; time += dtime) {
    f.time = (double) time;
    if (do_fields) {
	render_rectangle(&f, image, cps[0].width, field_even, channels, 0);
	f.time += 0.5;
	render_rectangle(&f, image, cps[0].width, field_odd, channels, 0);
    } else {
	render_rectangle(&f, image, cps[0].width, field_both, channels, 0);
    }
    if (1) {
	FILE *fp;
	if (getenv("out"))
	    fname = getenv("out");
	else
	    sprintf(fname, "%s%04d.%s", prefix, time, format);
	fp = fopen(fname, "wb");
	if (NULL == fp) {
	    perror(fname);
	    exit(1);
	}
	if (!strcmp(format, "png")) {
	   write_png(fp, image, cps[0].width, cps[0].height);
	} else if (!strcmp(format, "jpg")) {
	    write_jpeg(fp, image, cps[0].width, cps[0].height);
	} else {
	    fprintf(fp, "P6\n");
	    fprintf(fp, "%d %d\n255\n", cps[0].width, cps[0].height);
	    fwrite(image, 1, 3 * cps[0].width * cps[0].height, fp);
	}
	fclose(fp);
    }
  }
  return 0;
}
